presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java

```diff
@@ -405,22 +404,22 @@ private static void generateRowEqualsRowMethod(
         Parameter leftPage = arg("leftPage", Page.class);
         Parameter rightPosition = arg("rightPosition", int.class);
         Parameter rightPage = arg("rightPage", Page.class);
+        Parameter pagesHashChannels = arg("pagesHashChannels", int[].class);
         MethodDefinition rowEqualsRowMethod = classDefinition.declareMethod(
                 a(PUBLIC),
                 "rowEqualsRow",
                 type(boolean.class),
                 leftPosition,
                 leftPage,
                 rightPosition,
-                rightPage);
+                rightPage,
+                pagesHashChannels);
 
-        Scope compilerContext = rowEqualsRowMethod.getScope();
         for (int index = 0; index < joinChannelTypes.size(); index++) {
             BytecodeExpression type = constantType(callSiteBinder, joinChannelTypes.get(index));
 
-            BytecodeExpression leftBlock = leftPage.invoke("getBlock", Block.class, constantInt(index));
-
-            BytecodeExpression rightBlock = rightPage.invoke("getBlock", Block.class, constantInt(index));
+            BytecodeExpression leftBlock = leftPage.invoke("getBlock", Block.class, pagesHashChannels.getElement(index));
+            BytecodeExpression rightBlock = rightPage.invoke("getBlock", Block.class, pagesHashChannels.getElement(index));
```

Comment:
This design will not work for compilation.  The compiler generates something like this:
```java
return 
  BIGINT.equalsTo(leftPage.getBlock(0), rightPage.getBlock(0) &&
  VARCHAR.equalsTo(leftPage.getBlock(1), rightPage.getBlock(1) &&
  DOUBLE.equalsTo(leftPage.getBlock(2), rightPage.getBlock(1);
```
This is done so the VM can inline the equality code into the generated code, because it always knows the columns are types that are being checked at compile time.  This code will generate something like this:
```java
return 
  BIGINT.equalsTo(leftPage.getBlock(channels[0]), rightPage.getBlock(channels[0]) &&
  VARCHAR.equalsTo(leftPage.getBlock(channels[1]), rightPage.getBlock(channels[1]) &&
  DOUBLE.equalsTo(leftPage.getBlock(channels[2]), rightPage.getBlock(channels[2]);
```
I believe the only way to efficiently do this is to generate the old code and the new `IngoreNulls` versions